feat(common): make `HttpParameterCodec` injectable

johnny-mh opened the pull request:
You can inject custom `HttpParameterCodec` to not use default parameter encoder.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Specific API server can't accept character. '+' so we need to encode that.

https://github.com/angular/angular/blob/5298b2bda34a8766b28c8425e447f94598b23901/packages/common/http/src/params.ts#L64

But angular's standard encoder decode that.

So developer needs to add custom encoder every http method invokes. and it's very annoying.

Related Issue Numbers: #19710, #23157 #18261, #18719, #11058

## What is the new behavior?

User can provide default custom http parameter codec.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```


## Other information

Johanneke commented:
This is great, and it's exactly what we need for our application!

Johanneke commented on packages/common/http/src/params.ts:
```diff
@@ -13,12 +13,12 @@
-export interface HttpParameterCodec {
-  encodeKey(key: string): string;
-  encodeValue(value: string): string;
+export abstract class HttpParameterCodec {
```
Not sure why this is now an abstract class instead of an interface?

johnny-mh replied:
I think you are right. I'll refactor that during weekend :)

johnny-mh committed:
feat(common): make `HttpParameterCodec` injectable

You can inject custom HttpParameterCodec to not use default
parameter encoder in entire API requests.

PR Close #32598

gund commented:
This would be very useful.
Is there any timeline update when we may expect this to be merged?

Splaktar commented:
Please rebase this and update the goldens (as CircleCI should suggest when you push after rebasing).

johnny-mh commented:
@Splaktar Thank you for letting me know!

LinBoLen commented:
when use get request in angular. the date string format as `yyyy-MM-dd'T'HH:mm:ssX` which present as '2020-01-01T12:12:12+0100'. the `+` in date string will not be escaped. server will receive space after decode url

JounQin commented:
It's 2021, any new??

mcalmus commented:
Relates to #11058

johnny-mh commented:
@Splaktar I rebased the code a couple hours ago. but some builds pending until now. could I know reason for this?

gkalpak commented:
@johnny-mh, it seems like a glitch in the GitHub-->CircleCI communication. I closed and reopened the PR to trigger another event for CircleCI and the CI jobs are running now :smiley:

johnny-mh commented:
@Splaktar Thank you. 2 tests are failed but they are not seems to be related with the pull request.

One of them is `//packages/compiler-cli/ngcc/test:integration` error and others below.

I would appreciate it if you could send me some information.

```
Chrome Mobile 72.0.3626 (Android 0.0.0) NgZone long stack trace coalescing shouldCoalesceEventChangeDetection = true, shouldCoalesceRunChangeDetection = false should only emit onMicroTaskEmpty once within the same event loop for ngZone.run in onMicrotaskEmpty subscription FAILED
	Expected $.length = 4 to equal 5.
	Expected $[4] = undefined to equal 'microTask empty'.
	Error: Expected $.length = 4 to equal 5.
	Expected $[4] = undefined to equal 'microTask empty'.
	    at <Jasmine>
	    at packages/core/test/zone/ng_zone_spec.ts:1066:27 <- dist/all/@angular/core/test/zone/ng_zone_spec.js:911:57
```

gkalpak commented:
These were flakes apparently. I restarted the jobs and they passed 🤷

johnny-mh commented:
@gkalpak Thanks for helping me get the build through 😁 Could you possibly letting me know why 3 pending builds in checks?

petebacondarwin commented:
The CI builds are all now passing.
We need to get together with @alxhub, since he has an alternative approach in a PR, to decide the best way forward here.

atscott commented:
closing as superseded by #37385
